flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java

```diff
@@ -285,94 +310,222 @@ public Status process() throws EventDeliveryException {
+            @Override
+            public Dataset<GenericRecord> run() {
+              return Datasets.load(datasetUri);
+            }
+          });
+
+      DatasetDescriptor descriptor = view.getDataset().getDescriptor();
+      String formatName = descriptor.getFormat().getName();
+      Preconditions.checkArgument(allowedFormats().contains(formatName),
+          "Unsupported format: " + formatName);
+
+      Schema newSchema = descriptor.getSchema();
+      if (datasetSchema == null || !newSchema.equals(datasetSchema)) {
+        this.datasetSchema = descriptor.getSchema();
+        // dataset schema has changed, create a new parser
+        parser = parserFactory.newParser(datasetSchema, context);
+      }
+
+      this.reuseEntity = !("parquet".equals(formatName));
+      // Commit on batch if kite.commitOnBatch is true and the format isn't
+      // parquet
+      this.commitOnBatch = context.getBoolean(CONFIG_COMMIT_ON_BATCH,
+          DEFAULT_COMMIT_ON_BATCH) && !("parquet".equals(formatName));
+      this.datasetName = view.getDataset().getName();
+
+      return view.newWriter();
+    } catch (DatasetNotFoundException ex) {
+      throw new EventDeliveryException("Dataset " + datasetUri + " not found."
+          + " The dataset must be created before Flume can write to it.", ex);
+    } catch (RuntimeException ex) {
+      throw new EventDeliveryException("Error trying to open a new"
+          + " writer for dataset " + datasetUri + ": "
+          + ex.getLocalizedMessage(), ex);
     }
   }
 
-  private static Schema schema(Event event) throws EventDeliveryException {
-    Map<String, String> headers = event.getHeaders();
-    String schemaURL = headers.get(
-        DatasetSinkConstants.AVRO_SCHEMA_URL_HEADER);
+  /**
+   * Return true if the sink should roll the writer.
+   * 
+   * Currently, this is based on time since the last roll.
+   * 
+   * @return True if and only if the sink should roll the writer
+   */
+  private boolean shouldRoll() {
+    long currentTimeMillis = System.currentTimeMillis();
+    long elapsedTimeSeconds = TimeUnit.MILLISECONDS.toSeconds(
+        currentTimeMillis - lastRolledMs);
+
+    LOG.debug("Current time: {}, lastRolled: {}, diff: {} sec",
+        new Object[] {currentTimeMillis, lastRolledMs, elapsedTimeSeconds});
+
+    return elapsedTimeSeconds >= rollIntervalS;
+  }
+
+  /**
+   * Roll the dataset writer. If there's an error during the roll, the
+   * transaction will be rolled back.
+   * 
+   * @param closeWriter If true, the current writer is closed
+   * @param createNewWriter If true, a new writer is created
+   * @throws EventDeliveryException There was an error during the roll which
+   *                                means we can't guarantee previously written
+   *                                records were persisted.
+   */
+  @VisibleForTesting
+  void rollWriter(boolean closeWriter, boolean createNewWriter)
```

Comment:
Uses of this method aren't very clear because it [accepts two booleans](http://qt-project.org/wiki/API_Design_Principles#a6396e641a952364172f8879fe3eb4c1). It would be better to break it into separate methods (close and create) that can be invoked separately, or to get rid of at least one argument by renaming it to "rollWriterInternal" with the boolean and adding "closeAndRollWriter" and "rollWithoutClose" or similar methods that are easier to read.